third_party/move/move-vm/runtime/src/session.rs

```diff
@@ -13,19 +13,53 @@ use move_binary_format::{
 };
 use move_core_types::{
     account_address::AccountAddress,
-    effects::{ChangeSet, Event},
+    effects::{AccountChangeSet, ChangeSet, Changes, Event},
     gas_algebra::NumBytes,
     identifier::IdentStr,
     language_storage::{ModuleId, TypeTag},
     value::MoveTypeLayout,
+    vm_status::StatusCode,
 };
 use move_vm_types::{
     gas::GasMeter,
     loaded_data::runtime_types::{CachedStructIndex, StructType, Type},
-    values::GlobalValue,
+    values::{GlobalValue, Value},
 };
 use std::{borrow::Borrow, sync::Arc};
 
+fn serialize_changes(changes: Changes<Vec<u8>, (Value, MoveTypeLayout)>) -> VMResult<ChangeSet> {
+    let mut serialized_change_set = ChangeSet::new();
+    for (addr, account_change_set) in changes.into_inner() {
+        let mut serialized_account_change_set = AccountChangeSet::new();
+
+        // Currently all modules are serialized.
+        let (modules, resources) = account_change_set.into_inner();
+        for (identifier, blob_op) in modules {
+            serialized_account_change_set
+                .add_module_op(identifier, blob_op)
+                .expect("All identifiers are unique.");
+        }
+
+        // Resource stored as (value, layout) pairs. Serialize them as well.
+        for (tag, resource_op) in resources {
+            let blob_op = resource_op.and_then(|(value, layout)| {
```

Comment:
if we had a function serialize_with_layout for now instead of defining and_then, could we re-use it here and in adapter session finish as well? (later we may have to generalize, but maybe simplifies this PR)